File: libs/test-utils.js
/**
 * Yoimiya Test Utilities - Simple testing and validation helpers
 * 
 * This module provides convenient utilities for testing proofs, verifications,
 * and batch operations without needing to understand the internal implementation.
 */

const { generateTestSrs, proveTest, aggregateProofs } = require('../bindings/nodejs');


class YoimiyaTester {
    /**
     * Initialize tester with SRS
     * @param {number} maxDegree - Maximum polynomial degree
     */
    constructor(maxDegree = 2048) {
        this.srs = generateTestSrs(maxDegree);
        this.maxDegree = maxDegree;
        this.testResults = [];
    }

    /**
     * Test basic proof generation and verification
     * @param {number} numConstraints - Number of circuit constraints
     * @param {Array<bigint>} witness - Optional witness data
     * @returns {Object} Test result with timing and status
     */
    testSimpleProof(numConstraints = 100, witness = null) {
        if (!witness) {
            witness = [1n, 2n, 3n, 4n];
        }

        const result = {
            test: 'simple_proof',
            constraints: numConstraints,
            status: 'FAILED'
        };

        try {
            // Measure proof generation time
            let start = process.hrtime.bigint();
            const proof = proveTest(numConstraints, witness, this.srs);
            const proveTime = Number(process.hrtime.bigint() - start) / 1_000_000; // Convert to ms

            // Measure verification time
            start = process.hrtime.bigint();
            const valid = proof.verify(this.srs);
            const verifyTime = Number(process.hrtime.bigint() - start) / 1_000_000;

            result.status = valid ? 'PASSED' : 'FAILED';
            result.prove_ms = Math.round(proveTime * 10000) / 10000;
            result.verify_ms = Math.round(verifyTime * 10000) / 10000;
            result.proof_valid = valid;

        } catch (error) {
            result.error = error.message;
        }

        this.testResults.push(result);
        return result;
    }

    /**
     * Test proof aggregation and batch verification
     * @param {number} numProofs - Number of proofs to aggregate
     * @param {number} constraintsPerProof - Constraints per individual proof
     * @param {Array<bigint>} witness - Optional witness data
     * @returns {Object} Test result with timing and status
     */
    testBatchAggregation(numProofs = 5, constraintsPerProof = 100, witness = null) {
        if (!witness) {
            witness = [1n, 2n, 3n, 4n];
        }

        const result = {
            test: 'batch_aggregation',
            num_proofs: numProofs,
            constraints_per_proof: constraintsPerProof,
            status: 'FAILED'
        };

        try {
            // Generate multiple proofs
            const proofs = [];
            for (let i = 0; i < numProofs; i++) {
                const proof = proveTest(constraintsPerProof, witness, this.srs);
                proofs.push(proof);
            }

            // Measure aggregation time
            let start = process.hrtime.bigint();
            const batchProof = aggregateProofs(proofs, this.srs);
            const aggregateTime = Number(process.hrtime.bigint() - start) / 1_000_000;

            // Measure batch verification time
            start = process.hrtime.bigint();
            const valid = batchProof.verify(this.srs);
            const verifyTime = Number(process.hrtime.bigint() - start) / 1_000_000;

            result.status = valid ? 'PASSED' : 'FAILED';
            result.aggregate_ms = Math.round(aggregateTime * 10000) / 10000;
            result.batch_verify_ms = Math.round(verifyTime * 10000) / 10000;
            result.batch_valid = valid;

        } catch (error) {
            result.error = error.message;
        }

        this.testResults.push(result);
        return result;
    }

    /**
     * Test proof generation across different constraint sizes
     * @param {Array<number>} constraintSizes - List of constraint counts to test
     * @returns {Array<Object>} Test results for each size
     */
    testScalability(constraintSizes = null) {
        if (!constraintSizes) {
            constraintSizes = [100, 500, 1000, 2000];
        }

        const results = [];
        for (const size of constraintSizes) {
            const result = this.testSimpleProof(size, [1n, 2n, 3n, 4n]);
            results.push(result);
        }

        return results;
    }

    /**
     * Test aggregation with different batch sizes
     * @param {Array<number>} batchSizes - List of batch sizes to test
     * @param {number} constraintsPerProof - Constraints per individual proof
     * @returns {Array<Object>} Test results for each batch size
     */
    testBatchSizes(batchSizes = null, constraintsPerProof = 100) {
        if (!batchSizes) {
            batchSizes = [2, 5, 10, 20];
        }

        const results = [];
        for (const batchSize of batchSizes) {
            const result = this.testBatchAggregation(batchSize, constraintsPerProof);
            results.push(result);
        }

        return results;
    }

    /**
     * Run comprehensive test suite
     * @returns {Object} Summary of all tests
     */
    async runFullTestSuite() {
        console.log('Running Yoimiya Test Suite...');
        console.log('='.repeat(60));

        // Test 1: Simple proofs
        console.log('\n[1/4] Testing simple proof generation and verification...');
        const simpleResults = this.testScalability([100, 500, 1000, 2000]);

        // Test 2: Batch aggregation
        console.log('[2/4] Testing batch aggregation...');
        const batchResults = this.testBatchSizes([2, 5, 10]);

        // Test 3: Large batch
        console.log('[3/4] Testing large batch (100 proofs)...');
        const largeBatch = this.testBatchAggregation(100, 100);

        // Test 4: Stress test
        console.log('[4/4] Testing high-constraint proof...');
        const stressTest = this.testSimpleProof(5000, [1n, 2n, 3n, 4n]);

        // Print summary
        this._printSummary();

        return {
            simple: simpleResults,
            batches: batchResults,
            large_batch: largeBatch,
            stress: stressTest
        };
    }

    /**
     * Print test results summary
     * @private
     */
    _printSummary() {
        console.log('\n' + '='.repeat(60));
        console.log('TEST SUMMARY');
        console.log('='.repeat(60));

        const passed = this.testResults.filter(r => r.status === 'PASSED').length;
        const failed = this.testResults.filter(r => r.status === 'FAILED').length;

        console.log(`\nTotal tests: ${this.testResults.length}`);
        console.log(`✓ Passed: ${passed}`);
        console.log(`✗ Failed: ${failed}`);

        console.log('\nDetailed Results:');
        console.log('-'.repeat(60));
        for (const result of this.testResults) {
            const status = result.status === 'PASSED' ? '✓' : '✗';
            const testName = result.test || 'unknown';

            if (testName === 'simple_proof') {
                console.log(`${status} ${testName.padEnd(20)} | Constraints: ${String(result.constraints).padEnd(6)} | ` +
                    `Prove: ${result.prove_ms}ms | Verify: ${result.verify_ms}ms`);
            } else if (testName === 'batch_aggregation') {
                console.log(`${status} ${testName.padEnd(20)} | Proofs: ${String(result.num_proofs).padEnd(6)} | ` +
                    `Aggregate: ${result.aggregate_ms}ms | Verify: ${result.batch_verify_ms}ms`);
            }
        }
    }
}

/**
 * Quick sanity check test
 */
async function quickTest() {
    console.log('Running quick sanity check...');
    const tester = new YoimiyaTester(1024);
    const result = tester.testSimpleProof(100);

    if (result.status === 'PASSED') {
        console.log(`✓ SDK is working! Proof generation: ${result.prove_ms}ms, ` +
            `Verification: ${result.verify_ms}ms`);
    } else {
        console.log('✗ SDK test failed!');
        if (result.error) {
            console.log(`Error: ${result.error}`);
        }
    }

    return result;
}

module.exports = {
    YoimiyaTester,
    quickTest
};

File: libs/test-utils.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace Yoimiya.TestUtils
{
    /// <summary>
    /// High-level interface for testing Yoimiya SDK functionality
    /// Provides utilities for testing proofs, verifications, and batch operations
    /// </summary>
    public class YoimiyaTester
    {
        private readonly YoimiyaSdk.Srs _srs;
        private readonly int _maxDegree;
        private readonly List<TestResult> _testResults;

        /// <summary>
        /// Initialize tester with SRS
        /// </summary>
        public YoimiyaTester(int maxDegree = 2048)
        {
            _maxDegree = maxDegree;
            _srs = YoimiyaSdk.GenerateTestSrs(maxDegree);
            _testResults = new List<TestResult>();
        }

        /// <summary>
        /// Test basic proof generation and verification
        /// </summary>
        public TestResult TestSimpleProof(int numConstraints = 100, byte[] witness = null)
        {
            if (witness == null)
            {
                witness = new byte[] { 1, 2, 3, 4 };
            }

            var result = new TestResult
            {
                Test = "simple_proof",
                Constraints = numConstraints,
                Status = "FAILED"
            };

            try
            {
                // Measure proof generation time
                var sw = Stopwatch.StartNew();
                var proof = YoimiyaSdk.ProveTest(numConstraints, witness, _srs);
                sw.Stop();
                var proveMs = sw.Elapsed.TotalMilliseconds;

                // Measure verification time
                sw.Restart();
                var valid = proof.Verify(_srs);
                sw.Stop();
                var verifyMs = sw.Elapsed.TotalMilliseconds;

                result.Status = valid ? "PASSED" : "FAILED";
                result.ProveMs = Math.Round(proveMs, 4);
                result.VerifyMs = Math.Round(verifyMs, 4);
                result.ProofValid = valid;
            }
            catch (Exception ex)
            {
                result.Error = ex.Message;
            }

            _testResults.Add(result);
            return result;
        }

        /// <summary>
        /// Test proof aggregation and batch verification
        /// </summary>
        public TestResult TestBatchAggregation(int numProofs = 5, int constraintsPerProof = 100, byte[] witness = null)
        {
            if (witness == null)
            {
                witness = new byte[] { 1, 2, 3, 4 };
            }

            var result = new TestResult
            {
                Test = "batch_aggregation",
                NumProofs = numProofs,
                ConstraintsPerProof = constraintsPerProof,
                Status = "FAILED"
            };

            try
            {
                // Generate multiple proofs
                var proofs = new List<YoimiyaSdk.Proof>();
                for (int i = 0; i < numProofs; i++)
                {
                    var proof = YoimiyaSdk.ProveTest(constraintsPerProof, witness, _srs);
                    proofs.Add(proof);
                }

                // Measure aggregation time
                var sw = Stopwatch.StartNew();
                var batchProof = YoimiyaSdk.AggregateProofs(proofs.ToArray(), _srs);
                sw.Stop();
                var aggregateMs = sw.Elapsed.TotalMilliseconds;

                // Measure batch verification time
                sw.Restart();
                var valid = batchProof.Verify(_srs);
                sw.Stop();
                var verifyMs = sw.Elapsed.TotalMilliseconds;

                result.Status = valid ? "PASSED" : "FAILED";
                result.AggregateMs = Math.Round(aggregateMs, 4);
                result.BatchVerifyMs = Math.Round(verifyMs, 4);
                result.BatchValid = valid;
            }
            catch (Exception ex)
            {
                result.Error = ex.Message;
            }

            _testResults.Add(result);
            return result;
        }

        /// <summary>
        /// Test proof generation across different constraint sizes
        /// </summary>
        public List<TestResult> TestScalability(int[] constraintSizes = null)
        {
            if (constraintSizes == null)
            {
                constraintSizes = new[] { 100, 500, 1000, 2000 };
            }

            var results = new List<TestResult>();
            foreach (var size in constraintSizes)
            {
                var result = TestSimpleProof(size, new byte[] { 1, 2, 3, 4 });
                results.Add(result);
            }

            return results;
        }

        /// <summary>
        /// Test proof generation with large constraint counts (up to 1M)
        /// Useful for developers to understand performance at scale and identify
        /// optimal parameters for their use case.
        /// </summary>
        public List<TestResult> TestLargeConstraints(int[] constraintSizes = null)
        {
            if (constraintSizes == null)
            {
                constraintSizes = new[] { 10_000, 50_000, 100_000, 250_000, 500_000, 1_000_000 };
            }

            Console.WriteLine("\nTesting Large Constraint Sizes (up to 1M)...");
            Console.WriteLine(new string('-', 60));

            var results = new List<TestResult>();
            foreach (var size in constraintSizes)
            {
                Console.Write($"  Testing {size:N0} constraints... ");
                var result = TestSimpleProof(size, new byte[] { 1, 2, 3, 4 });
                if (result.Status == "PASSED")
                {
                    Console.WriteLine($"✓ Prove: {result.ProveMs}ms, Verify: {result.VerifyMs}ms");
                }
                else
                {
                    Console.WriteLine($"✗ Failed: {result.Error ?? "Unknown error"}");
                }
                results.Add(result);
            }

            return results;
        }

        /// <summary>
        /// Test aggregation with different batch sizes
        /// </summary>
        public List<TestResult> TestBatchSizes(int[] batchSizes = null, int constraintsPerProof = 100)
        {
            if (batchSizes == null)
            {
                batchSizes = new[] { 2, 5, 10, 20 };
            }

            var results = new List<TestResult>();
            foreach (var batchSize in batchSizes)
            {
                var result = TestBatchAggregation(batchSize, constraintsPerProof);
                results.Add(result);
            }

            return results;
        }

        /// <summary>
        /// Run comprehensive test suite
        /// </summary>
        public TestSuiteResult RunFullTestSuite()
        {
            Console.WriteLine("Running Yoimiya Test Suite...");
            Console.WriteLine(new string('=', 60));

            // Test 1: Simple proofs
            Console.WriteLine("\n[1/4] Testing simple proof generation and verification...");
            var simpleResults = TestScalability(new[] { 100, 500, 1000, 2000 });

            // Test 2: Batch aggregation
            Console.WriteLine("[2/4] Testing batch aggregation...");
            var batchResults = TestBatchSizes(new[] { 2, 5, 10 });

            // Test 3: Large batch
            Console.WriteLine("[3/4] Testing large batch (100 proofs)...");
            var largeBatch = TestBatchAggregation(100, 100);

            // Test 4: Stress test
            Console.WriteLine("[4/4] Testing high-constraint proof...");
            var stressTest = TestSimpleProof(5000, new byte[] { 1, 2, 3, 4 });

            // Print summary
            PrintSummary();

            return new TestSuiteResult
            {
                Simple = simpleResults,
                Batches = batchResults,
                LargeBatch = largeBatch,
                Stress = stressTest
            };
        }

        /// <summary>
        /// Print test results summary
        /// </summary>
        private void PrintSummary()
        {
            Console.WriteLine("\n" + new string('=', 60));
            Console.WriteLine("TEST SUMMARY");
            Console.WriteLine(new string('=', 60));

            var passed = _testResults.Count(r => r.Status == "PASSED");
            var failed = _testResults.Count(r => r.Status == "FAILED");

            Console.WriteLine($"\nTotal tests: {_testResults.Count}");
            Console.WriteLine($"✓ Passed: {passed}");
            Console.WriteLine($"✗ Failed: {failed}");

            Console.WriteLine("\nDetailed Results:");
            Console.WriteLine(new string('-', 60));
            foreach (var result in _testResults)
            {
                var status = result.Status == "PASSED" ? "✓" : "✗";
                var testName = result.Test ?? "unknown";

                if (testName == "simple_proof")
                {
                    Console.WriteLine($"{status} {testName.PadEnd(20)} | Constraints: {result.Constraints.ToString().PadEnd(6)} | " +
                        $"Prove: {result.ProveMs}ms | Verify: {result.VerifyMs}ms");
                }
                else if (testName == "batch_aggregation")
                {
                    Console.WriteLine($"{status} {testName.PadEnd(20)} | Proofs: {result.NumProofs.ToString().PadEnd(6)} | " +
                        $"Aggregate: {result.AggregateMs}ms | Verify: {result.BatchVerifyMs}ms");
                }
            }
        }
    }

    /// <summary>
    /// Individual test result
    /// </summary>
    public class TestResult
    {
        public string Test { get; set; }
        public string Status { get; set; }
        public int Constraints { get; set; }
        public int NumProofs { get; set; }
        public int ConstraintsPerProof { get; set; }
        public double ProveMs { get; set; }
        public double VerifyMs { get; set; }
        public double AggregateMs { get; set; }
        public double BatchVerifyMs { get; set; }
        public bool ProofValid { get; set; }
        public bool BatchValid { get; set; }
        public string Error { get; set; }
    }

    /// <summary>
    /// Full test suite result
    /// </summary>
    public class TestSuiteResult
    {
        public List<TestResult> Simple { get; set; }
        public List<TestResult> Batches { get; set; }
        public TestResult LargeBatch { get; set; }
        public TestResult Stress { get; set; }
    }

    /// <summary>
    /// Quick test utility
    /// </summary>
    public static class QuickTest
    {
        public static TestResult Run()
        {
            Console.WriteLine("Running quick sanity check...");
            var tester = new YoimiyaTester(1024);
            var result = tester.TestSimpleProof(100);

            if (result.Status == "PASSED")
            {
                Console.WriteLine($"✓ SDK is working! Proof generation: {result.ProveMs}ms, " +
                    $"Verification: {result.VerifyMs}ms");
            }
            else
            {
                Console.WriteLine("✗ SDK test failed!");
                if (!string.IsNullOrEmpty(result.Error))
                {
                    Console.WriteLine($"Error: {result.Error}");
                }
            }

            return result;
        }
    }
}
